pkg/autodiscovery/listeners/snmp.go

```diff
@@ -0,0 +1,351 @@
+	err := config.Datadog.UnmarshalKey("snmp_listener", &snmpConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &SNMPListener{
+		services: map[string]Service{},
+		devices:  map[string][]string{},
+		stop:     make(chan bool),
+		config:   snmpConfig,
+	}, nil
+}
+
+// Listen periodically refreshes devices
+func (l *SNMPListener) Listen(newSvc chan<- Service, delSvc chan<- Service) {
+	// setup the I/O channels
+	l.newService = newSvc
+	l.delService = delSvc
+
+	go l.checkDevices()
+}
+
+func (l *SNMPListener) loadCache(config util.SNMPConfig, adIdentifier string, cacheKey string) {
+	cacheValue, err := persistentcache.Read(cacheKey)
+	if err != nil {
+		log.Errorf("Couldn't read cache for %s: %s", cacheKey, err)
+		return
+	}
+	if cacheValue == "" {
+		return
+	}
+	var devices []net.IP
+	err = json.Unmarshal([]byte(cacheValue), &devices)
+	if err != nil {
+		log.Errorf("Couldn't unmarshal cache for %s: %s", cacheKey, err)
+		return
+	}
+	for _, deviceIP := range devices {
+		entityID := config.Digest(deviceIP.String())
+		l.createService(deviceIP.String(), adIdentifier, entityID)
+	}
+}
+
+func writeCache(cacheKey string, devices []string) {
+	cacheValue, err := json.Marshal(devices)
+	if err != nil {
+		log.Errorf("Couldn't marshal cache: %s", err)
+		return
+	}
+	err = persistentcache.Write(cacheKey, string(cacheValue))
+	if err != nil {
+		log.Errorf("Couldn't write cache: %s", err)
+	}
+}
+
+// Don't make it a method, to be overridden in tests
+var worker = func(l *SNMPListener, jobs <-chan snmpSubnet) {
+	for {
+		select {
+		case <-l.stop:
+			log.Debug("Stopping SNMP worker")
+			return
+		case subnet := <-jobs:
+			log.Debugf("Handling IP %s", subnet.currentIP.String())
+			l.checkDevice(subnet.adIdentifier, subnet.currentIP.String(), subnet.config, subnet.defaultParams)
+		default:
+		}
+	}
+}
+
+func (l *SNMPListener) checkDevice(adIdentifier string, deviceIP string, config util.SNMPConfig, defaultParams *gosnmp.GoSNMP) {
+	params := *defaultParams
+	params.Target = deviceIP
+	entityID := config.Digest(deviceIP)
+	err := params.Connect()
+	if err != nil {
+		log.Debugf("SNMP connect to %s error: %v", deviceIP, err)
+		l.deleteService(entityID)
+	} else {
+		defer params.Conn.Close()
+
+		oids := []string{"1.3.6.1.2.1.1.2.0"}
+		v, err := params.Get(oids)
+		if err != nil || len(v.Variables) < 1 || v.Variables[0].Value == nil {
+			log.Debugf("SNMP get to %s error: %v", deviceIP, err)
+			l.deleteService(entityID)
+		} else {
+			log.Debugf("SNMP get to %s success: %v", deviceIP, v.Variables[0].Value)
+			l.createService(deviceIP, adIdentifier, entityID)
+		}
+	}
+}
+
+func (l *SNMPListener) checkDevices() {
+	subnets := []snmpSubnet{}
+	for i, config := range l.config.Configs {
+		ipAddr, ipNet, err := net.ParseCIDR(config.Network)
+		if err != nil {
+			return
+		}
+
+		defaultParams, err := config.BuildSNMPParams()
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+		adIdentifier := fmt.Sprintf("snmp_%d", i)
```

Comment:
Is it good enough? Using the config order is probably not great, shouid I use the config has instead?

Reply:
> Using the config order is probably not great

Indeed, I'm not familiar with SNMP and the SNMP check, but in AD; AD identifiers are identifiers of the targeted Service (e.g container name, container image).
Can you think of something similar for SNMP? it should identify the IP or the subset of IPs targeted by one check instance. Also the config provider must be able to get it and add it to the config template - AD configs cannot be resolved it the AD identifiers from the listener and the config provider don't match!

Reply:
Right, I think the config hash will make more sense then. Will change it.